### PR TITLE
Addition of nll_loss and cross_entropy to tensor.py

### PR DESCRIPTION
**Pull request by airpods69**
* tensor: added nll_loss and cross_entropy
* test_ops: added test for nll_loss and test for cross_entropy
* 
This PR adds negative_log_likelihood and cross_entropy to Tensor. #3891 and #5247 

Eg:
For `negative_log_likelihood`,
```py
import tinygrad
data = tinygrad.Tensor([[1, 2, 4]])
target = [2]
print(data.nll_loss(target).numpy())
```

For `cross_entropy`,
```py
import tinygrad
data = tinygrad.Tensor([[1, 2, 4]])
target = [2]
print(data.cross_entropy(target).numpy())
```

======= Issues to Resolve =======
The issue to resolve lies within the `test/test_ops.py` where the test fails for the added functions (`test_nll_loss` and `test_cross_entropy`) when `forward_only` is set to `False`. The error returned is that there is no grad available in the Tensor.

The returned values from the implemented loss functions matches that of `torch`.

Conversation to refer to: https://discord.com/channels/1068976834382925865/1070745817025106080/1266685261618872350  

Logs: https://gist.github.com/nighteous/2992fc80703f8e15a55d44b3455d9620

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -2856,6 +2856,15 @@ def one_hot(self, num_classes:int) -> Tensor:
...
     return (self[..., None] == Tensor.arange(num_classes, requires_grad=False, device=self.device)).where(1, 0)
 
+  def nll_loss(self, target:Tensor) -> Tensor:
+    """
+    Returns the nll_loss for self with target.
+    """
+    return -self.gather(1, target.view(1, target.shape[0]).T).mean()
+
+  def cross_entropy(self, target:Tensor) -> Tensor:
+    return self.log_softmax().nll_loss(target).mean()
```
Comment (wozeparrot) [inline tinygrad/tensor.py:2866]: missing docs

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -2856,6 +2856,15 @@ def one_hot(self, num_classes:int) -> Tensor:
     """
     return (self[..., None] == Tensor.arange(num_classes, requires_grad=False, device=self.device)).where(1, 0)
 
+  def nll_loss(self, target:Tensor) -> Tensor:
+    """
+    Returns the nll_loss for self with target.
+    """
+    return -self.gather(1, target.view(1, target.shape[0]).T).mean()
```
Comment (wozeparrot) [inline tinygrad/tensor.py:2863]: docs doesn't follow standard format

File: test/test_ops.py [hunk 3]
```diff
@@ -1971,6 +1971,19 @@ def test_one_hot(self):
     helper_test_op([], lambda: torch.nn.functional.one_hot(torch.tensor(data), 8).type(torch.int32),
                        lambda: Tensor(data).one_hot(8), forward_only=True)
 
+  def test_nll_loss(self):
+    helper_test_op([(4, 4), (4)], lambda x,y: torch.nn.functional.nll_loss(x,torch.clip(y, 0).type(torch.long), reduction = 'mean'),
+                                  lambda x,y: x.nll_loss(y.clip(0).cast(dtypes.long)), forward_only=True)
+
+  def test_cross_entropy(self):
+    helper_test_op([(4, 4), (4)], lambda x,y: torch.nn.functional.cross_entropy(x,torch.clip(y, 0).type(torch.long), reduction = 'mean'),
+                                       lambda x,y: x.cross_entropy(y.clip(0,1)), forward_only=True)
```
Comment (wozeparrot) [inline test/test_ops.py:1980]: no tests for backward?

**Review by wozeparrot — CHANGES_REQUESTED** [review]
ideally api would match torch, missing grad could be from `gather`

File: test/test_ops.py (hunk 3 above)
Comment (airpods69) [reply test/test_ops.py:1980]: it is one of the issues to resolved and I wanted to see if it actually worked without format or not. (Mentioned in the top comment)

**Comment by airpods69** [comment]
> ideally api would match torch, missing grad could be from `gather`

yes that is what I am trying to figure out how to calculate `nll_loss` without gather. Any directions would be great cause I have been stuck on this for a while.

**Comment by geohot** [comment]
I'm interested in this, still finishing it up?

**Comment by airpods69** [comment]
> I'm interested in this, still finishing it up?

Finishing up would be more like being closer to the goal. The gather function is not differentiable cause of which backward tasks fail but from my understanding, how `nll_loss` works, you need gather to collect and average out the losses at the indices [the labels].
Any comments or directions and I will complete it by tomorrow too if it makes sense.

If you want to continue from here, then please be my guest.
